bigtable-dataflow-parent/bigtable-dataflow-import/src/main/java/com/google/cloud/bigtable/dataflowimport/HBaseExportJob.java

```diff
@@ -0,0 +1,303 @@
+    final Scan scan = new Scan()
+        .setStartRow(options.getStartRow().getBytes())
+        .setStopRow(options.getStopRow().getBytes())
+        .setMaxVersions(options.getMaxVersions());
+
+    if (!options.getFilter().isEmpty()) {
+      scan.setFilter(new ParseFilter().parseFilterString(options.getFilter()));
+    }
+
+    final String bigtableProject = Objects.firstNonNull(options.getBigtableProjectId(), options.getProject());
+
+    final CloudBigtableScanConfiguration scanConfig = new CloudBigtableScanConfiguration.Builder()
+        .withProjectId(bigtableProject)
+        .withInstanceId(options.getBigtableInstanceId())
+        .withTableId(options.getBigtableTableId())
+        .withScan(scan)
+        .build();
+
+    final BoundedSource<Result> source = CloudBigtableIO.read(scanConfig);
+
+
+    // Build the pipeline
+    final Pipeline pipeline = Pipeline.create(options);
+
+    // Coders
+    final SerializableCoder<BoundedSource<Result>> sourceCoder = SerializableCoder.of(new TypeDescriptor<BoundedSource<Result>>() {});
+    final KvCoder<String, BoundedSource<Result>> kvCoder = KvCoder.of(StringUtf8Coder.of(), sourceCoder);
+
+    // Pipeline
+    pipeline
+        // Initialize with the source as the only element.
+        // Note that we aren't reading the source's elements yet
+        .apply("Initialize", Create.of(source).withCoder(sourceCoder))
+        // Split the source into a bunch of shards
+        // Note that we aren't reading the source's elements yet
+        .apply("Shard", ParDo.of(new ShardScans<Result>())).setCoder(kvCoder)
+        // Make sure that each shard can be handled a different worker
+        .apply("Fanout", Reshuffle.<String, BoundedSource<Result>>of())
+        // Now, read the actual rows and write out each shard's elements into a file
+        .apply("Write", ParDo.of(new WriteResultsToSeq(options.getDestination())));
+
+    return pipeline;
+  }
+
+  /**
+   * Split Sources into multiple shards. It treats the Sources themselves as elements. It will output
+   * a KV of a random unique key and a sharded Source. It will intentionally randomize the order of shards
+   * to avoid hotspotting later on.
+   */
+  static class ShardScans<T> extends DoFn<BoundedSource<T>, KV<String, BoundedSource<T>>> {
+    private static final long DESIRED_SHARD_SIZE = 1024 * 1024 * 1024;
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      final BoundedSource<T> source = c.element();
+
+      final List<? extends BoundedSource<T>> boundedSources = source
+          .splitIntoBundles(DESIRED_SHARD_SIZE, c.getPipelineOptions());
+
+      LOG.info("Split into {} shards", boundedSources.size());
+
+      // randomize order to make sure that scans are distributed across tablets
+      Collections.shuffle(boundedSources);
+
+      // emit each shard with a random uuid, note that uuids are used instead of counters to prevent misleading the
```

Comment:
The keys allow for the bundles keep their filenames stable. If a shard fails to write to gcs, we want it to retry using the same name. Keeping the filename as a key makes that very easy.